modules/pallets/messaging-fees/src/impls.rs

```diff
@@ -0,0 +1,228 @@
+						Request::Get(_) => 0,
+					},
+					IncentivizedMessage::Response(res) => match res {
+						Response::Post(post) => post.response.len() as u32,
+						Response::Get(get) => get
+							.values
+							.iter()
+							.filter_map(|storage_val| storage_val.value.as_ref())
+							.map(|bytes| bytes.len())
+							.sum::<usize>() as u32,
+					},
+				})
+				.sum();
+
+			TotalBytesProcessed::<T>::mutate(|total| {
+				*total = total.saturating_add(bytes_processed)
+			});
+			let current_total_bytes = TotalBytesProcessed::<T>::get();
+
+			if let Some(host_params) =
+				pallet_ismp_host_executive::HostParams::<T>::get(&state_machine)
+			{
+				let per_byte_fee = match host_params {
+					EvmHostParam(evm_host_param) => {
+						let fee = evm_host_param
+							.per_byte_fees
+							.iter()
+							.find(|fee| {
+								let hashed_chain_id = keccak_256(&state_machine.encode());
+								fee.state_id == H256(hashed_chain_id)
+							})
+							.map(|fee| fee.per_byte_fee)
+							.unwrap_or(evm_host_param.default_per_byte_fee);
+
+						fee
+					},
+					SubstrateHostParam(VersionedHostParams::V1(substrate_params)) => {
+						let fee = substrate_params
+							.per_byte_fees
+							.get(&state_machine)
+							.cloned()
+							.unwrap_or(substrate_params.default_per_byte_fee);
+						let fee_u128: u128 = fee.into();
+						U256::from(fee_u128)
+					},
+				};
+
+				let dollar_cost = per_byte_fee.saturating_mul(U256::from(bytes_processed));
+				let dollar_cost: u128 =
+					dollar_cost.try_into().map_err(|_| Error::<T>::CalculationOverflow)?;
+
+				let bridge_usd_price = T::PriceOracle::get_price()
+					.map_err(|_| Error::<T>::ErrorInPriceConversion)?;
+
+				let base_reward_in_token =
+					bridge_usd_price.saturating_mul(dollar_cost.saturated_into());
+
+				let target_message_size = T::TargetMessageSize::get();
+
+				if current_total_bytes <= target_message_size {
+					let reward_amount = Self::calculate_reward(
+						current_total_bytes,
+						target_message_size,
+						base_reward_in_token.into(),
+					)?;
+
+					T::Currency::transfer(
+						&T::TreasuryAccount::get().into_account_truncating(),
+						&relayer,
+						reward_amount.saturated_into(),
+						Preservation::Expendable,
+					)
+					.map_err(|_| Error::<T>::RewardTransferFailed)?;
+					Self::deposit_event(Event::FeeRewarded {
+						relayer: relayer.clone(),
+						amount: reward_amount.saturated_into(),
+					});
+				} else {
+					T::Currency::transfer(
+						&relayer,
+						&T::TreasuryAccount::get().into_account_truncating(),
+						base_reward_in_token,
+						Preservation::Expendable,
+					)
+					.map_err(|_| Error::<T>::RewardTransferFailed)?;
+
+					Self::deposit_event(Event::FeePaid {
+						relayer: relayer.clone(),
+						amount: base_reward_in_token.into(),
+					});
+				}
+			}
+		}
+
+		Ok(())
+	}
+
+	/// A curve for calculating reward
+	/// Reward=BaseReward×((TargetSize−TotalBytes)/TargetSize)^2
+	pub fn calculate_reward(
```

Comment:
Write a test that can visualize how this reward function works when incentivizing up to 1 megabyte